### PR TITLE
feat: use simple filter to prune memtable 

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Use simple filter to avoid constructing record batch. Build arrays from value is costly and will use lots of memory. Memory comparison for after and before (`before` is killed manually). 

image

From @zhongzc's report it can occupies ~60GB:
image

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Comment by v0y4g3r** [comment]
Do we have to fallback to physical expr evaluator when given predicates cannot be converted to simple evaluator?

File: src/mito2/src/memtable/time_series.rs [hunk 1]
```diff
@@ -369,13 +362,46 @@ struct Iter {
...
+        codec: Arc<McmpRowCodec>,
+    ) -> Self {
+        let simple_filters = predicate
+            .map(|p| {
+                p.exprs()
+                    .iter()
+                    .filter_map(|f| SimpleFilterEvaluator::try_new(f.df_expr()))
+                    .collect::<Vec<_>>()
+            })
+            .unwrap_or_default();
```
Comment (v0y4g3r) [inline src/mito2/src/memtable/time_series.rs:390]: Do we have to fallback to physical expr evaluator when given predicates cannot be converted to simple evaluator?

Comment (waynexia) [reply src/mito2/src/memtable/time_series.rs:390]: Predicates here are not required to be accurate. We can ignore them if evaluating on physical expr is expansive. And the filter behavior of parquet and memtable are aligned now.

**Review by v0y4g3r — APPROVED** [review]
LGTM

File: src/mito2/src/memtable/time_series.rs [hunk 2]
```diff
@@ -305,24 +299,23 @@ impl SeriesSet {
...
-            last_key: None,
-            predicate: physical_exprs,
-            pk_schema: primary_key_schema,
-            primary_key_builders,
-            codec: self.codec.clone(),
-            metrics: Metrics::default(),
-        }
+            predicate,
+            primary_key_schema,
+            // primary_key_builders,
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:315]:
```suggestion
```

File: src/mito2/src/memtable/time_series.rs [hunk 3]
```diff
@@ -305,24 +299,23 @@ impl SeriesSet {
 
     /// Iterates all series in [SeriesSet].
     fn iter_series(&self, projection: HashSet<ColumnId>, predicate: Option<Predicate>) -> Iter {
-        let (primary_key_builders, primary_key_schema) =
-            primary_key_builders(&self.region_metadata, 1);
-
-        let physical_exprs: Vec<_> = predicate
-            .and_then(|p| p.to_physical_exprs(&primary_key_schema).ok())
-            .unwrap_or_default();
+        let (_, primary_key_schema) = primary_key_builders(&self.region_metadata, 1);
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:302]: Implement a function `primary_key_schema()` to do it?

File: src/mito2/src/memtable/time_series.rs [hunk 4]
```diff
@@ -369,13 +362,46 @@ struct Iter {
     series: Arc<SeriesRwLockMap>,
     projection: HashSet<ColumnId>,
     last_key: Option<Vec<u8>>,
-    predicate: Vec<Arc<dyn PhysicalExpr>>,
+    predicate: Vec<SimpleFilterEvaluator>,
     pk_schema: arrow::datatypes::SchemaRef,
-    primary_key_builders: Vec<Box<dyn MutableVector>>,
+    // primary_key_builders: Vec<Box<dyn MutableVector>>,
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:367]:
```suggestion
```

File: src/mito2/src/memtable/time_series.rs [hunk 5]
```diff
@@ -415,7 +441,8 @@ impl Iterator for Iter {
                     &self.codec,
                     primary_key.as_slice(),
                     &mut series,
-                    &mut self.primary_key_builders,
+                    // &mut self.primary_key_builders,
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:444]:
```suggestion
```

File: src/mito2/src/memtable/time_series.rs [hunk 6]
```diff
@@ -446,88 +473,49 @@ fn prune_primary_key(
     codec: &Arc<McmpRowCodec>,
     pk: &[u8],
     series: &mut Series,
-    builders: &mut [Box<dyn MutableVector>],
+    // builders: &mut [Box<dyn MutableVector>],
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:476]:
```suggestion
```

**Review by evenyag — APPROVED** [review]
Rest LGTM. Some temporary codes could be removed.

**Pull request merged**